exporters/prometheus/src/test/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverterTest.java

```diff
@@ -0,0 +1,257 @@
+    assertThat(matcher.matches()).isTrue();
+    assertThat(matcher.group("help")).isEqualTo(expectedHelp);
+    assertThat(matcher.group("type")).isEqualTo(expectedType);
+    // Note: Summaries and histograms produce output which matches METRIC_NAME_PATTERN multiple
+    // times. The pattern ends up matching against the first.
+    assertThat(matcher.group("metricName")).isEqualTo(expectedMetricName);
+  }
+
+  private static Stream<Arguments> metricMetadataArgs() {
+    // TODO (jack-berg): delete "Previously metricName was .." comments before merging, update
+    // comments to reflect new logic
+    return Stream.of(
+        // special case for gauge
+        // Previously metricName was "sample_ratio"
+        Arguments.of(
+            createSampleMetricData("sample", "1", MetricDataType.LONG_GAUGE),
+            "sample_ratio gauge",
+            "sample_ratio description",
+            "sample_ratio"),
+        // special case for gauge with drop - metric unit should match "1" to be converted to
+        // "ratio"
+        // Previously metricName was "sample"
+        Arguments.of(
+            createSampleMetricData("sample", "1{dropped}", MetricDataType.LONG_GAUGE),
+            "sample_ratio gauge",
+            "sample_ratio description",
+            "sample_ratio"),
+        // Gauge without "1" as unit
+        // Previously metricName was "sample_unit"
+        Arguments.of(
+            createSampleMetricData("sample", "unit", MetricDataType.LONG_GAUGE),
+            "sample_unit gauge",
+            "sample_unit description",
+            "sample_unit"),
+        // special case with counter
+        // Previously metricName was "sample_unit_total"
+        Arguments.of(
+            createSampleMetricData("sample", "unit", MetricDataType.LONG_SUM),
+            "sample_unit_total counter",
+            "sample_unit_total description",
+            "sample_unit_total"),
+        // special case unit "1", but not gauge - "1" is dropped
+        // Previously metricName was "sample_total"
+        Arguments.of(
+            createSampleMetricData("sample", "1", MetricDataType.LONG_SUM),
+            "sample_ratio_total counter",
+            "sample_ratio_total description",
+            "sample_ratio_total"),
+        // units expressed as numbers other than 1 are retained
+        // Previously metricName was "sample_2_total"
+        Arguments.of(
+            createSampleMetricData("sample", "2", MetricDataType.LONG_SUM),
+            "sample_2_total counter",
+            "sample_2_total description",
+            "sample_2_total"),
+        // metric name with unsupported characters
+        // Previously metricName was "s_ple_percent_per_minute"
+        Arguments.of(
+            createSampleMetricData("s%%ple", "%/min", MetricDataType.SUMMARY),
+            "s__ple_percent_per_minute summary",
+            "s__ple_percent_per_minute description",
+            "s__ple_percent_per_minute_count"),
```

Comment:
Dang.. it doesn't look like the prometheus client name sanitization does this. I could add a `name.replaceAll("__", "_")` after sanitizing the metric name:

```
    String name = sanitizeMetricName(metricData.getName());
    String help = metricData.getDescription();
    Unit unit = PrometheusUnitsHelper.convertUnit(metricData.getUnit());
    if (unit != null && !name.endsWith(unit.toString())) {
      // Need to re-sanitize metric name since unit may contain illegal characters
      name = sanitizeMetricName(name + "_" + unit);
    }
    name = name.replaceAll("__", "_");

    return new MetricMetadata(name, help, unit);
```
But we're starting to allocate a lot of strings. I wonder why the otel spec talks about this while the prometheus client doesn't seem to care? 🤔

Reply:
In Prometheus metric names with double underscores are allowed. Not sure why the spec says double underscores should be stripped.

However, if we want to remove double underscores and avoid String allocations we could do something like

```java
    while (name.contains("__")) {
      name.replace("__", "_");
    }
```
`name.contains("__")` will return `false` in most cases, so the loop will almost never be executed.